Mathlib/RingTheory/DedekindDomain/FiniteAdeleRing.lean

```diff
@@ -364,6 +370,120 @@ instance : Algebra (R_hat R K) (FiniteAdeleRing R K) where
   commutes' _ _ := mul_comm _ _
   smul_def' r x := rfl
 
+instance : CoeFun (FiniteAdeleRing R K)
+    (fun _ ↦ ∀ (v : HeightOneSpectrum R), adicCompletion K v) where
+  coe a v := a.1 v
+
+open scoped algebraMap in
+lemma exists_finiteIntegralAdele_iff (a : FiniteAdeleRing R K) : (∃ c : R_hat R K,
+    a = c) ↔ ∀ (v : HeightOneSpectrum R), a v ∈ adicCompletionIntegers K v :=
+  ⟨by rintro ⟨c, rfl⟩ v; exact (c v).2, fun h ↦ ⟨fun v ↦ ⟨a v, h v⟩, rfl⟩⟩
+
+section Topology
+
+open Classical
+
+open nonZeroDivisors
+
+open scoped algebraMap -- coercion from R to FiniteAdeleRing R K
+
+open scoped DiscreteValuation
+
+open Multiplicative Additive
+
+variable {R K} in
+lemma clear_local_denominator (v : HeightOneSpectrum R)
```

Comment:
This is a general lemma right? Nothing adele-specific. So I guess it can move to the file on adic completions?

Reply:
Yes, R is global and K_v is local and we're just saying that we can clear a local denominator with a global element. I'll move it.

Reply:
Moved to the adic completion file.